lxd/device/disk.go

```diff
@@ -349,6 +349,42 @@ func (d *disk) validateConfig(instConf instance.ConfigReader) error {
 					return fmt.Errorf("Custom filesystem volumes require a path to be defined")
 				}
 			}
+
+			// Extract initial configuration from the profile and validate them against appropriate
+			// storage driver. Currently initial configuration is only applicable to root disk devices.
+			initialConfig := make(map[string]string)
+			for k, v := range d.config {
+				prefix, newKey, found := strings.Cut(k, "initial.")
+				if found && prefix == "" {
+					initialConfig[newKey] = v
+				}
+			}
+
+			if len(initialConfig) > 0 {
+				if !shared.IsRootDiskDevice(d.config) {
+					return fmt.Errorf("Non-root disk device cannot contain initial.* configuration")
+				}
+
+				volumeType, err := storagePools.InstanceTypeToVolumeType(d.inst.Type())
+				if err != nil {
+					return err
+				}
+
+				// Create temporary volume definition.
+				vol := storageDrivers.NewVolume(
+					d.pool.Driver(),
+					d.pool.Name(),
+					volumeType,
+					storagePools.InstanceContentType(d.inst),
+					d.name,
+					initialConfig,
+					d.pool.Driver().Config())
+
+				err = d.pool.Driver().ValidateVolume(vol, false)
+				if err != nil {
+					return fmt.Errorf("Invalid initial device configuration: %v", err)
+				}
+			}
```

Comment:
@tomponline  I think this is the last bit that has to be improved.

Currently, expanded initial configuration is checked against the appropriate storage driver. This causes VMs to be blocked from creating if, for example, `initial.zfs.block_mode` is set for root device in profile.

However, we cannot use `err = d.pool.Driver().ValidateVolume(vol, true)` (skip unused keys) as then this will cause options that are directly provided to the instance to be simply ignored. For example:
```sh
# In such case, this option will not be applied, but no error will be thrown.
lxc launch ubuntu:22.04 c1 -d root,some_invalid_option=true
```

So, I suggest, that we extract initial configuration from expeneded devices and local devices separately. We will then ignore unused keys for expended devices. For local devices we do not allow any unused keys.

Any thoughts?

Reply:
This approach won't work, because if the option is provided on the instance level, profile and instance options are merged.

```
$ lxc profile show test
config: {}
description: ""
devices:
  eth0:
    nictype: bridged
    parent: lxdbr0
    type: nic
  root:
    initial.zfs.block_mode: "true"
    path: /
    pool: default
    type: disk
name: test
used_by: []

$ lxc launch $image c2 -p test -d root,initial.block.filesystem=btrfs
Creating c2
Error: Failed instance creation: Failed creating instance record: Failed initialising instance: Failed add validation for device "root": Invalid initial device configuration: Invalid option for volume "root" option "zfs.block_mode"
```